<a href="https://colab.research.google.com/github/JonRivera/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_t_tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [0]:
### YOUR WORK HERE
import pandas as pd
import numpy as np
from scipy import stats
columns =['symboling', 'normalized-losses','make','fuel-type','aspiration','num-of-doors',
'body-style','drive-wheels','engine-location','wheel-base','length','width','height','curb-weight',
'engine-type','num-of-cylinders','engine-size','fuel-system','bore','stroke','compression-ratio','horsepower',
'peak-rpm','city-mpg','highway-mpg','price']
df = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data", names = columns)

In [0]:
print(df.head(),df.tail(),df.shape)

   symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]      symboling normalized-losses   make  ... city-mpg highway-mpg  price
200         -1                95  volvo  ...       23          28  16845
201         -1                95  volvo  ...       19          25  19045
202         -1                95  volvo  ...       18          23  21485
203         -1                95  volvo  ...       26          27  22470
204         -1                95  volvo  ...       19          25  22625

[5 rows x 26 columns] (205, 26)


In [0]:
#Clean Data, Put headers, replace ? with NaN values
df = df.replace({"?":np.NaN}) 
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [0]:
### YOUR WORK HERE
#Sample of 10 cars with a random state of 30 
df_sample = df.sample(10, random_state=30)
df_sample


symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056
103          0               108         nissan  ...       19          25  13499
26           1               148          dodge  ...       31          38   7609
20           0                81      chevrolet  ...       38          43   6575
155          0                91         toyota  ...       27          32   8778
185          2                94     volkswagen  ...       27          34   8195

[10 rows x 26 columns]

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [151]:
### YOUR WORK HERE
# Check for descriptive statistics
# Just from this summary we can see we have an average of 32.8 mpg for highway
df_sample['highway-mpg'].mean()

32.8

1.  Null Hypothesis: Cars sold typically have a fuel efficiency of 35mpg.
2. Alternative Hypothesis: Cars are sold at some other fuel efficency that is not 35mpg.

3. Confidence level: The confidence level will be  95% or we need a .5 p value or greater to accept the Null hypothesis.

In [0]:
# Run a 1 Sample T test
# Here our expected for for null hypothesis mean is 35

stats.ttest_1samp(df_sample['highway-mpg'],35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

4. The P value is .36, which is greater that 0.05
5. Conclusion: Based on a statistic of -0.96 and a p value of 0.36, I fail  to reject the null hypothesis.


Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [150]:
df_sample2 = df.sample(100, random_state=30)
df_sample2['highway-mpg'].mean()

30.64

1.  Null Hypothesis: Cars sold typically have a fuel efficiency of 35mpg.
2. Alternative Hypothesis: Cars are sold at some other fuel efficency that is not 35mpg.

3. Confidence level: The confidence level will be  95% or we need a .5 p value or greater to accept the Null hypothesis.

In [0]:
#Run A 1 sample T Test for 100 cars in the lot
# From descriptive statistics we get an average mean of 30.64
stats.ttest_1samp(df_sample2['highway-mpg'],35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

4. The  P value of 3.00 * e-09, which is less than 0.5
5. Conclusion: Based on a statistic of -6.518 and a p value of 3.00 * e-09 I reject the null hypothesis that the cars sold have a 35mpg fuel efficiency. 

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [0]:
### YOUR WORK HERE
columns1 = ['party', 'handicapped-infants','water-project-cost-sharing',
          'adoption-of-the-budget-resolution','physician-fee-freeze', 'el-salvador-aid',
          'religious-groups-in-schools','anti-satellite-test-ban',
          'aid-to-nicaraguan-contras','mx-missile','immigration',
           'synfuels-corporation-cutback','education-spending',
           'superfund-right-to-sue','crime','duty-free-exports',
           'export-administration-act-south-africa']
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names = columns1)
print(df.shape)
print(df.head(), df.tail())


(435, 17)
        party  ... export-administration-act-south-africa
0  republican  ...                                      y
1  republican  ...                                      ?
2    democrat  ...                                      n
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]           party  ... export-administration-act-south-africa
430  republican  ...                                      y
431    democrat  ...                                      y
432  republican  ...                                      y
433  republican  ...                                      y
434  republican  ...                                      n

[5 rows x 17 columns]


In [0]:
### YOUR WORK HERE
#Convert ? to nana, and y,n votes into binary numbers of either 0 or 1
df = df.replace({'?':np.NaN, 'y':1, 'n':0})

#check
print(df.head(), df.isnull().sum(), df['party'].value_counts())

        party  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns] party                                       0
handicapped-infants                        12
water-project-cost-sharing                 48
adoption-of-the-budget-resolution          11
physician-fee-freeze                       11
el-salvador-aid                            15
religious-groups-in-schools                11
anti-satellite-test-ban                    14
aid-to-nicaraguan-contras                  15
mx-missile                                 22
immigration                                 7
synfuels-corporation-cutback               21
education-spending                         31
superfund-right-to-sue   

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [0]:
dem = df[df['party'] =='democrat']

rep = df[df['party'] == 'republican'] 

In [0]:
dem.describe()

handicapped-infants  ...  export-administration-act-south-africa
count           258.000000  ...                              185.000000
mean              0.604651  ...                                0.935135
std               0.489876  ...                                0.246956
min               0.000000  ...                                0.000000
25%               0.000000  ...                                1.000000
50%               1.000000  ...                                1.000000
75%               1.000000  ...                                1.000000
max               1.000000  ...                                1.000000

[8 rows x 16 columns]

In [0]:
rep.describe()

handicapped-infants  ...  export-administration-act-south-africa
count           165.000000  ...                              146.000000
mean              0.187879  ...                                0.657534
std               0.391804  ...                                0.476168
min               0.000000  ...                                0.000000
25%               0.000000  ...                                0.000000
50%               0.000000  ...                                1.000000
75%               0.000000  ...                                1.000000
max               1.000000  ...                                1.000000

[8 rows x 16 columns]

Looking at the numerical description of both parties we can see that democrats highly favored adoption of the budget resolution, having a mean of 89% yes votes.
And we can also see that republicans are against it, having a mean of 13%, of the ones who voted yed. This is the Issue i will use for problem 1.

1. Null Hypothesis: Both parties support the adoption of the budget resolution bill at a similar percentage.
2. Alternative : Democrats support the adoption of the budget resolution bill more.
%
3. Confidence Level: 95%


In [0]:
#Run a two sample t test
stats.ttest_ind(dem['adoption-of-the-budget-resolution'], rep['adoption-of-the-budget-resolution'],nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

4. The p values is 1.994262314074344e-177)
5. Conclusion: Based on a statistic of 23.21 (positive) and a p values of 1.61* e-18 I reject the null hypothesis. Instead I suggest the alternative, that the democrats support the bill more.

The numerical summaries also tell us that republicans highly favored physician fee-freeze, having an average mean of 99%. While democrats were highly against it, having an average mean of 5.4% of yes votes. This is the issue I will use for problem 2.



1. Hypothesis: Both parties support the physician fee-free bill.
2. Alternative: Republicans support the physician fee-freeze bill more.
3. Confidence level = 95%


In [0]:
# Run a two sample test 
stats.ttest_ind(dem['physician-fee-freeze'], rep['physician-fee-freeze'],nan_policy='omit')

Ttest_indResult(statistic=-49.36708157301406, pvalue=1.994262314074344e-177)

4. The p values is 1.994262314074344e-177
5. Conclusion: Baed on a t statistic of -49.36 (negative suggesting more republicans support) and a p value of 1.99 e-177, I reject the null hypothesis and suggest the alternative, which is that republicans support the bill more.



An issue that both republicans seem to agree on is the water-project-cost-sharing; where the average mean for democrats was 50% and and the average of yes votes for rupublicans was 51%. This is the issue I will use for problem 3.

1. Null Hypothesis: Democrats and Republicans support the water-project cost sharing bill at a same rate.

2. The alternative, is that both parties  accept the bill at different rates.
3. Confidence level 95%

In [0]:
print([stats.ttest_ind(dem['water-project-cost-sharing'], rep['water-project-cost-sharing'],nan_policy='omit')])

[Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)]


4. P value of .93
5. Based on a T statisitic of 0.089 and a p value of .929 I suggest the null hypthesis is true and both republicans and democrats support the bill
at a similar percentage.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [0]:

all_sample_tests = []
bills = list(df)[1:]

def two_ttest(x):
  x = stats.ttest_ind(dem[x], rep[x],nan_policy='omit')
  return x

for bill in bills:
  all_sample_tests.append(two_ttest(bill))
all_sample_tests



[Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18),
 Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485),
 Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77),
 Ttest_indResult(statistic=-49.36708157301406, pvalue=1.994262314074344e-177),
 Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68),
 Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20),
 Ttest_indResult(statistic=12.526187929077842, pvalue=8.521033017443867e-31),
 Ttest_indResult(statistic=18.052093200819733, pvalue=2.82471841372357e-54),
 Ttest_indResult(statistic=16.437503268542994, pvalue=5.03079265310811e-47),
 Ttest_indResult(statistic=-1.7359117329695164, pvalue=0.08330248490425066),
 Ttest_indResult(statistic=8.293603989407588, pvalue=1.5759322301054064e-15),
 Ttest_indResult(statistic=-20.500685724563073, pvalue=1.8834203990450192e-64),
 Ttest_indResult(statistic=-13.51064251060933, pvalue=1.227858170



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [0]:
### YOUR WORK HERE


### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [0]:
### YOUR WORK HERE